<!DOCTYPE html>


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>I/O &mdash; dms2dfe 0.1.0 documentation</title>
    
    <link rel="stylesheet" href="_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/bootswatch-3.3.6/united/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="_static/bootstrap-sphinx.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="_static/js/jquery-1.11.0.min.js"></script>
    <script type="text/javascript" src="_static/js/jquery-fix.js"></script>
    <script type="text/javascript" src="_static/bootstrap-3.3.6/js/bootstrap.min.js"></script>
    <script type="text/javascript" src="_static/bootstrap-sphinx.js"></script>
    <link rel="top" title="dms2dfe 0.1.0 documentation" href="index.html" />
    <link rel="next" title="Example" href="6test.html" />
    <link rel="prev" title="Programs" href="3programs.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head>
  <body role="document">

  <div id="navbar" class="navbar navbar-default navbar-fixed-top">
    <div class="container">
      <div class="navbar-header">
        <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
        <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
          <span class="icon-bar"></span>
        </button>
        <a class="navbar-brand" href="index.html">
          dms2dfe</a>
        <span class="navbar-text navbar-version pull-left"><b>0.1.0</b></span>
      </div>

        <div class="collapse navbar-collapse nav-collapse">
          <ul class="nav navbar-nav">
            
                <li><a href="1installation.html">Installation</a></li>
                <li><a href="2usage.html">Usage</a></li>
                <li><a href="3programs.html">Programs</a></li>
                <li><a href="#">I/O</a></li>
                <li><a href="6test.html">Example</a></li>
            
            
              <li class="dropdown globaltoc-container">
  <a role="button"
     id="dLabelGlobalToc"
     data-toggle="dropdown"
     data-target="#"
     href="index.html">Site <b class="caret"></b></a>
  <ul class="dropdown-menu globaltoc"
      role="menu"
      aria-labelledby="dLabelGlobalToc"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="1installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="2usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="3programs.html">Programs</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="">I/O</a></li>
<li class="toctree-l1"><a class="reference internal" href="6test.html">Example</a></li>
<li class="toctree-l1"><a class="reference internal" href="5api.html">API</a></li>
</ul>
</ul>
</li>
              
                <li class="dropdown">
  <a role="button"
     id="dLabelLocalToc"
     data-toggle="dropdown"
     data-target="#"
     href="#">Page <b class="caret"></b></a>
  <ul class="dropdown-menu localtoc"
      role="menu"
      aria-labelledby="dLabelLocalToc"><ul>
<li><a class="reference internal" href="#">I/O</a><ul>
<li><a class="reference internal" href="#inputs-configuration">Inputs (configuration)</a></li>
<li><a class="reference internal" href="#format-of-info-configuration-file">Format of <cite>info</cite> configuration file</a></li>
<li><a class="reference internal" href="#formats-of-other-configuration-files">Formats of other configuration files</a></li>
<li><a class="reference internal" href="#outputs">Outputs</a></li>
<li><a class="reference internal" href="#format-of-data-lbl">Format of <cite>data_lbl</cite></a></li>
<li><a class="reference internal" href="#format-of-data-fit">Format of <cite>data_fit</cite></a></li>
<li><a class="reference internal" href="#format-of-data-comparison">Format of <cite>data_comparison</cite></a></li>
</ul>
</li>
</ul>
</ul>
</li>
              
            
            
            
            
            
          </ul>

          
            
<form class="navbar-form navbar-right" action="search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
          
        </div>
    </div>
  </div>

<div class="container">
  <div class="row">
    <div class="col-md-12 content">
      
  <div class="section" id="i-o">
<span id="io"></span><h1>I/O<a class="headerlink" href="#i-o" title="Permalink to this headline">¶</a></h1>
<p>Details about inputs and outputs of <a class="reference internal" href="index.html#dms2dfe"><span>dms2dfe</span></a>.</p>
<div class="section" id="inputs-configuration">
<h2>Inputs (configuration)<a class="headerlink" href="#inputs-configuration" title="Permalink to this headline">¶</a></h2>
<p>Inputs of dms2dfe are located in <cite>prj_dh/cfg/</cite>.
They are .csv files. Namely</p>
<div class="highlight-text"><div class="highlight"><pre>prj_dh (project directory)
|-- cfg
|   |-- info        : variables for a particular analysis eg. number of cores, host organism etc.
|   |-- lbls        : names of samples and corresponding paths to files.
|   |-- fit         : pairs of input (before selection) and selected (after selection) samples.
|   |-- comparison  : pairs of control (eg. untreated) and test (eg. treated) samples to be compared.
|   |-- repli       : replicates to be averaged
|   |-- feats       : custom features.
|   `-- barcodes    : information about deplexing barcoded .fastq files.
</pre></div>
</div>
</div>
<div class="section" id="format-of-info-configuration-file">
<h2>Format of <cite>info</cite> configuration file<a class="headerlink" href="#format-of-info-configuration-file" title="Permalink to this headline">¶</a></h2>
<p>This includes the variables for a particular analysis.</p>
<div class="highlight-text"><div class="highlight"><pre>prj_dh (project directory)
|-- cfg
|   |-- info
|   |   |-- column varname          : names of input variables
|   |   |-- column input            : inputs values
|   |   |-- column default          : default values
|   |   |-- column description      : details of values
|   |   |   |-- Path to reference fasta file
|   |   |   |-- Path to pdb file
|   |   |   |-- Host name for assigning codon table
|   |   |   |-- Number of cores to be used
|   |   |   |-- Alignment type [loc:local| glob:global]
|   |   |   |-- Cut off for frequency per mutant
|   |   |   |-- Type of normalization across samples
|   |   |   |   [wild: wrt wild type | syn : wrt synonymous mutations | none : fold change serves as fitness]
|   |   |   |-- Optional: residue numbers of active sites (space delimited) eg. 68&lt;SPACE&gt;192
|   |   |   |-- Optional: path to dssp module
|   |   |   |-- Optional: path to trimmomatic source (.jar) file
|   |   |   |-- Optional: path to bowtie2 source file
|   |   |   |-- Optional: path to samtools source file
|   |   |   `-- Optional: if reference sequence is concatamer
</pre></div>
</div>
<p><cite>info</cite> can be edited in multiple ways as describes in <a class="reference internal" href="2usage.html#usage"><span>Usage</span></a>.</p>
</div>
<div class="section" id="formats-of-other-configuration-files">
<h2>Formats of other configuration files<a class="headerlink" href="#formats-of-other-configuration-files" title="Permalink to this headline">¶</a></h2>
<p>In addition, other csv files in <cite>prj_dh/cfg/</cite> need to be filled.</p>
<div class="highlight-text"><div class="highlight"><pre>prj_dh (project directory)
|-- cfg
|   |-- lbls                        : names of samples and corresponding paths to files.
|   |   |-- column  varname         : sample names,
|   |   |-- column  fhs_1           : paths to input files (.fastq or .sam or .bam or .mut_cds)
|   |   `-- column  fhs_2           : paths to reverse (R2) of pair ended .fastq files.
|   |-- fit                         : pairs of input (before selection) and selected (after selection) samples.
|   |   |-- column  unsel           : names of input (unselected) samples,
|   |   `-- columns sel_1,sel_2,..  : names of selected samples
|   |-- comparison                  : pairs of control (eg. untreated) and test (eg. treated) samples to be compared.
|   |   |-- column  ctrl            : names of control samples,
|   |   `-- columns test_1,test_2,..: names of test samples
|   |-- repli                       : replicates to be averaged.
|   |   |-- column  varname         : names of averages samples
|   |   `-- columns replicate_1,replicate_2,.. : names of replicates
|   |-- feats                       : custom features.
|   |   |-- column  aasi            : amino acid number,
|   |   `-- columns feat_1,feat_2,..: features (eg. conservation score).
|   |-- barcodes                    : information about deplexing barcoded .fastq files.
|   |   |-- column  fastq_R1_fh     : paths to forward (R1) fastq files,
|   |   |-- column  fastq_R2_fh     : paths to reverse (R2) fastq files,
|   |   |-- column  barcode_R1      : sequence of forward (R1) barcode,
|   |   |-- column  barcode_R2      : sequence of reverse (R2) barcode,
|   |   `-- column  fastq_fn        : file names of output files.
</pre></div>
</div>
<p>For templates please have a look at test datasets located in <cite>dms2dfe/test_dataset/</cite>.</p>
</div>
<div class="section" id="outputs">
<h2>Outputs<a class="headerlink" href="#outputs" title="Permalink to this headline">¶</a></h2>
<p>With a successful run of <cite>dms2dfe</cite> following output files are created.
Below is structure of output files generated.</p>
<div class="highlight-text"><div class="highlight"><pre>prj_dh (project directory)
|-- data_lbl        : frequencies of mutants
|   |-- aas         : amino acid level
|   `-- cds         : codon level
|-- data_fit        : fitness of mutants
|   |-- aas         : amino acid level
|   `-- cds         : codon level
|-- data_comparison : comparison of fitness of mutants between test (eg. treated) and control (eg. untreated).
|   |-- aas         : amino acid level
|   `-- cds         : codon level
|-- plots
|   |-- aas         : amino acid level
|   `-- cds         : codon level
|-- data_ml         : data used for generating classification model.
|-- data_feats              : features generated by ana0_getfeats.
</pre></div>
</div>
</div>
<div class="section" id="format-of-data-lbl">
<h2>Format of <cite>data_lbl</cite><a class="headerlink" href="#format-of-data-lbl" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text"><div class="highlight"><pre>prj_dh (project directory)
|-- data_lbl                                : frequencies of mutants
|   `-- aas/cds
|       |-- column  ref                      : Wild-type residue/nucleotide
|       |-- column  mut                      : Mutated residue/nucleotide
|       |-- column  mutids                   : ID of mutant &lt;ref&gt;&lt;index&gt;&lt;mut&gt;
|       |-- columns NiA,NiAcut,NiAcutlog    : raw, thresholded(cut) and log of depth for all(A) mutants
|       |-- columns NiS,NiScut,NiScutlog    : raw, thresholded(cut) and log of depth for synonymous(S) mutants
|       `-- columns NiN,NiNcut,NiNcutlog    : raw, thresholded(cut) and log of depth for non-synonymous(N) mutants
</pre></div>
</div>
</div>
<div class="section" id="format-of-data-fit">
<h2>Format of <cite>data_fit</cite><a class="headerlink" href="#format-of-data-fit" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text"><div class="highlight"><pre>prj_dh (project directory)
|-- data_fit                            : fitness of mutants
|   `-- aas/cds
|       |-- column  ref                  : Wild-type residue/nucleotide
|       |-- column  mut                  : Mutated residue/nucleotide
|       |-- column  mutids               : ID of mutant &lt;ref&gt;&lt;index&gt;&lt;mut&gt;
|       |-- columns NiAunsel,NiAsel     : log of depth for all(A) mutants in unselected(unsel) and selected(sel) pool
|       |-- columns NiSunsel,NiSsel     : log of depth for synonymous(S) mutants in unselected(unsel) and selected(sel) pool
|       |-- columns FCA,FiA             : Fold change(FC) and fitness(Fi) of all(A) mutants
|       |-- columns FCS,FiS             : Fold change(FC) and fitness(Fi) of synonymous(S) mutants
|       `-- column  class_fit           : class of fitness
|           |-- value = beneficial      : if fitness(Fi) &gt; 2
|           |-- value = neutral         : if fitness(Fi) &lt; 2 &amp; &gt; -2
|           |-- value = deleterious     : if fitness(Fi) &lt; -2
|           |-- value = killed          : if not detected in selected(sel) pool
|           `-- value = survived        : if not detected in unselected(unsel) pool
</pre></div>
</div>
</div>
<div class="section" id="format-of-data-comparison">
<h2>Format of <cite>data_comparison</cite><a class="headerlink" href="#format-of-data-comparison" title="Permalink to this headline">¶</a></h2>
<div class="highlight-text"><div class="highlight"><pre>prj_dh (project directory)
|-- data_comparison                     : comparison of fitness of mutants between test (eg. treated) and control (eg. untreated).
|   `-- aas/cds
|       |-- column ref                  : Wild-type residue/nucleotide
|       |-- column mut                  : Mutated residue/nucleotide
|       |-- column mutids               : ID of mutant &lt;ref&gt;&lt;index&gt;&lt;mut&gt;
|       |-- columns Fi_ctrl,Fi_test    : Fitness(Fi) of all(A) mutants in control and test respectively
|       |-- columns class_fit_ctrl,class_fit_test     : class of fitness in control and test respectively
|       `-- column class_comparison     : class of comparison
|           |-- value = positive        : if class_fit_ctrl != beneficial  &amp; class_fit_test == beneficial
|           |-- value = stabilizing     : if class_fit_ctrl != neutral     &amp; class_fit_test == neutral
|           |-- value = negative        : if class_fit_ctrl != deleterious &amp; class_fit_test == deleterious
|           `-- value = robust          : if class_fit_ctrl == class_fit_test
</pre></div>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
        <br/>
        
<div id="sourcelink">
  <a href="_sources/4io.txt"
     rel="nofollow">Source</a>
</div>
      
    </p>
    <p>
        &copy; Copyright 2016, Rohan Dandage.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.3.1.<br/>
    </p>
  </div>
</footer>
  </body>
</html>